Fix #5817

mjschwenne opened the pull request:
Ok, so first of all this fix was way faster than I thought it would be, thanks in part to my excessively detailed notes and blog posts from last summer on the topic (such as [this one](https://mjschwenne.github.io/2021/06/03/a-closer-look-at-held-karp.html)). The linear program that gets run as part of `direction_of_ascent` is checking for the existence of a set of non-negative constants. If the program is feasible, then the direction of ascent doesn't exist and we can stop trying to find it. 

So, I believe that `optimize.linprog` failing to find a solution to the program is fine; it just means that this set of constants doesn't exist and we have to keep looking. More information can be found on page 13 of [this paper](https://pubsonline.informs.org/doi/abs/10.1287/opre.18.6.1138) by Held and Karp titled "The Traveling-Salesman Problem and Minimum Spanning Trees". So in reality, just checking if the problem has a solution gives us all the information we need. Note that I any taking advantage of the default non-negative bounds that the `linprog` on the variables involved.  

Since linear programming is *not* something I claim any more than the most basic skill with (although I would be interesting in taking a class about it if given the chance) input from someone with a bit more expertise would help clam some lingering doubts that I just can't seem to smooth over in part because this particular internal function is so difficult to generate test cases for.

dschult commented:
Nice!  This fix didn't take long. 
Fixes #5817  
Just so I understand better -- 

1) it looks like you've changed from the `interior-point` method to a default method.  Does that have any implications?

2) You nicely switch `arb_count` to be tracked via an `enumerate` call.  But it looks like you left the increment statement for `arb_count` in the code. It doesn't affect anything I believe, but maybe that line should be removed too.

mjschwenne commented:
1. Actually I was always using the default method. The change to `interior-point` was done by @rossbar as a temporary fix to the problem since when the program fails, `interior-point` returns an array of all zeros while the new default method in 1.9 returns `None` which was the original source of the bug. As far as I know there are no implications with the changing methods. They all solve the same problem and the algorithm here basically just needs to check if the problem is feasible. I trust the scipy people to set the default to be the most performant and robust implementation that they have. 
2. Must have missed the line when I added the `enumerate` call. I have fixed it.

dschult reviewed (approved):
Thanks!!
